Allow debugging the stdout of the electron main process

Raynos opened the pull request:
When using `testcafe` to run tests against an electron app we do not have any visibility into the behavior of the electron main process.

If anything writes to stdout or stderr its ignored by default.

Raynos commented:
@AndreyBelym Any update on this PR ? 

Improving debugging would be great and the change is pretty small.

If you want me to change the behavior like using a command line flag instead of an environment variable then I can do that too.

AndreyBelym commented on src/index.js:
```diff
@@ -31,7 +31,7 @@ function startElectron (config, ports) {
     }
 
-    spawn(cmd, args, { stdio: 'ignore' });
+    spawn(cmd, args, { stdio: process.env.DEBUG ? 'inherit' : 'ignore' });
```
The `DEBUG` environment variable has some special meaning for many Node.js since it controls features of the [debug](https://www.npmjs.com/package/debug) module. The `testcafe` module uses it as well. I would like to use `pipe` in the `stdio` configuration and always send the piped output to the `debug` module. Something like:
```js
import debug from 'debug';

const debugLogger = debug('testcafe-browser-provider-electron');
const stdioLogger = debugLogger.extend('spawn').extend('stdio');

...

const process = spawn(cmd, args, { stdio: ['ignore', 'pipe', 'pipe'] });

for (const stream of [process.stdout, process.stderr])
    stream.on('data', data => stdioLogger(String(data)));

Raynos replied:
Willfix. I’ll add and configure debug

Raynos commented:
I addressed feedback; cc @AndreyBelym 

I looked at the source code of `testcafe` itself and it looks like the `debug(testcafe:${scope})` convention is used.

Here I am using `debug(testcafe:electron)` and for these specific output lines I am using `spawn:stdout` and spawn:stderr` as the suffix so that stdout & stderr can be differentiated.

By using `testcafe:electron:spawn` we can add other debug log lines in the future for `testcafe:electron:${some_feature}` 

By prefixing with `testcafe:` the upstream user can set `DEBUG=testcafe:*` when submitting a bug report to the DevExpress team and it will contain all debug logs related to `testcafe` incluidng first party browser providers like electron.

AndreyBelym commented on src/index.js:
```diff
@@ -15,6 +16,9 @@ import ERRORS from './errors';
 
 import testRunTracker from 'testcafe/lib/api/test-run-tracker';
 
+const DEBUG_LOGGER = debug('testcafe:electron');
```
It's a great idea to use `testcafe` as a top namespace, but I would like to use `browser-provider-electron` on the second level to be consistent with the package name.

Raynos replied:
willfix.

Raynos commented:
This time actually passing CI. I can squash these commits if you want.

AndreyBelym commented on src/index.js:
```diff
@@ -31,7 +35,14 @@ function startElectron (config, ports) {
-    spawn(cmd, args, { stdio: 'ignore' });
+    var proc = spawn(cmd, args, { stdio: ['ignore', 'pipe', 'pipe'] });
+
+    proc.stdout.on('data', (buf) => {
```
The TestCafe codestyle prefers writing arrow functions in a single line when it's possible. The data handlers are quite short, so we can rewrite them as one-liners.

Raynos replied:
willfix.

AndreyBelym commented on src/index.js:
```diff
@@ -31,7 +35,14 @@ function startElectron (config, ports) {
+    var proc = spawn(cmd, args, { stdio: ['ignore', 'pipe', 'pipe'] });
+
+    proc.stdout.on('data', (buf) => {
+        STDOUT_LOGGER(buf.toString().trim());
```
It comes from my personal experience, but I try to avoid processing debug output. I know, the output can be a bit ugly without `trim()`, but simplicity is better than beauty for debug logs. 

Also, `String(value)` can handle primitive values like numbers, `null` and `undefined` without exceptions, so it's better to use it for debugging since in JS world you get literally anything as a result and we don't want to ruin the app's process with an exception in the debug logger.

Raynos replied:
the `buf` value is part of the node core `child_process` API and its a `Buffer` ; converting it to a string should really be `buf.toString('utf8')` but `String(buf)` does work as well.

As for trim; I want to trim a trailing new line because the output has a trailing new line and the `debug` module also prints a new line.

This leads to double new line output; the same problem would happen if you did `console.log(buf.toString('utf8'))` instead of `process.stdout.write(buf)` ;

console.log adds a new line for you, `process.stdout.write` just writes buffers to stdout and your responsible for new lines.

Raynos commented:
Addressed feedback.

AndreyBelym commented on src/index.js:
```diff
@@ -31,7 +35,22 @@ function startElectron (config, ports) {
+    });
+    proc.stderr.on('data', (buf) => {
+        var str = String(buf);
+        if (str[str.length - 1] === '\n') {
```
We can use [lodash.trimEnd(str, '\n')`](https://lodash.com/docs/4.17.15#trimEnd) here.

Raynos commented:
@andreybelym

Thanks, addressed feedback again.

AndreyBelym commented:
Thank you again, I will publish a new version in a couple of hours 👍

Raynos commented:
@AndreyBelym The new version has not been published yet.

AndreyBelym commented:
I apologize for the delay. `testcafe-browser-provider-electron@0.0.14` contains this feature and should be available now.

Raynos commented:
Awesome thanks.
